### PR TITLE
docs(x/ecocredit): update ecocredit spec

### PR DESCRIPTION
**Pull request by ryanchristo**
## Description

Closes: #604

Update ecocredit specification to include marketplace functionality.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct `docs:` prefix in the PR title
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the [documentation writing guidelines](https://github.com/cosmos/cosmos-sdk/blob/master/docs/DOC_WRITING_GUIDELINES.md)
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct `docs:` prefix in the PR title
- [ ] confirmed all author checklist items have been addressed 
- [ ] confirmed that this PR only changes documentation
- [ ] reviewed content for consistency
- [ ] reviewed content for thoroughness
- [ ] reviewed content for spelling and grammar
- [ ] tested instructions (if applicable)

### CODE REVIEW

File: x/ecocredit/spec/03_messages.md [hunk 1]
```diff
@@ -80,3 +80,51 @@ Cancel removes a number of credits from the holder's account and also deducts th
 - `credits` must not be empty
 - `batch_denom` must be a valid batch denomination
 - `amount` must be positive
+
+## MsgSell
+
+`Sell` creates one or more sell orders. Sell orders can be created in batches
```
Comment (blushi) [inline x/ecocredit/spec/03_messages.md:86]:
```suggestion
`Sell` creates one or more sell orders. Sell orders can be created in batches.
```

File: x/ecocredit/spec/03_messages.md [hunk 2]
```diff
@@ -80,3 +80,51 @@ Cancel removes a number of credits from the holder's account and also deducts th
...
+### Validation:
+
+- `owner` must ba a valid address, and their signature must be present in the transaction
+- `batch_denom` must be a valid credit batch denom
+- `quantity` must be a positive decimal
+- `ask_price` must be a positive integer
+
+## MsgUpdateSellOrders
+
+`UpdateSellOrders` updates one or more sell orders...
```
Comment (blushi) [inline x/ecocredit/spec/03_messages.md:99]: why using "..."?

File: x/ecocredit/spec/03_messages.md [hunk 3]
```diff
@@ -80,3 +80,51 @@ Cancel removes a number of credits from the holder's account and also deducts th
...
+
+## MsgAllowAskDenom
+
+`AllowAskDenom` adds a denom and details about the denom...
+
++++ https://github.com/regen-network/regen-ledger/blob/a70af9b4d5da6cf065eb23232ba8020769249688/proto/regen/ecocredit/v1alpha1/tx.proto#L430-L444
+
+### Validation:
+
+- `root_address` must ba the address of the governance module, and its signature must be present in the transaction
```
Comment (blushi) [inline x/ecocredit/spec/03_messages.md:129]:
```suggestion
- `root_address` must be the address of the governance module
```

File: x/ecocredit/spec/03_messages.md (hunk 2 above)
Comment (ryanchristo) [reply x/ecocredit/spec/03_messages.md]: I was probably planning to write more and forgot. Updated. :+1:

**Comment by ryanchristo** [comment]
> Don't we also wanna add project related docs?

I can add project related docs. The pull request was not merged at the time but it is now.

**Comment by ryanchristo** [comment]
I actually think it might be best to follow up with project updates in a separate pull request. The original issue was to update the docs with marketplace functionality. I started making the updates but this pull request is quite large and I'm finding a few other items related to projects that need to be fixed (comments in proto files and state table prefixes). Would love to get this pull request through rather than adding more to it and then address project updates separately if that's ok.

**Review by blushi — APPROVED** [review]
Makes sense, let's tackle project docs update in a separate PR, is there already an issue for this?

**Comment by ryanchristo** [comment]
> Makes sense, let's tackle project docs update in a separate PR, is there already an issue for this?

Opened #699.

File: x/ecocredit/spec/01_concepts.md [hunk 4]
```diff
@@ -53,6 +53,22 @@ Tradable credits are credits that can be transferred by the owner to another acc
 
 Retired credits are credits that cannot be transferred between accounts nor can they be unretired. Retired credits are equivalent to burned tokens with the exception that retired credits are actively tracked after being retired. Retiring a credit implies that the holder of a credit is “claiming” the credit as an offset. Credits can be retired upon issuance, retired upon transfer, and retired by the owner of the credits. The retirement location is required upon retirement.
 
+## Order Book
```
Comment (robert-zaremba) [inline x/ecocredit/spec/01_concepts.md:56]: Maybe let's rename to: "Simple Order Book", and below in description highlight that:
* there is no automatic matching
* users have to manually take the orders

File: x/ecocredit/spec/06_tests.md [hunk 5]
```diff
@@ -281,3 +269,185 @@ If a user tries to cancel credits and the user provides an invalid batch denomin
...
+
+If a user tries to create a buy order and the user provides a bid price using an ask denom that is NOT allowed, then the transaction fails and the buy order is NOT processed.
+
+- GIVEN - user provides a bid price using an ask denom that is NOT allowed
+- WHEN - user tries to create a buy order
+- THEN - transaction fails, buy order is NOT processed
+
+## Allow Ask Denom
+
+If a user tries to add an ask denom and the user submits a governance proposal that is approved, then the transaction is successful and the ask denom is added.
```
Comment (robert-zaremba) [inline x/ecocredit/spec/06_tests.md:437]: do we need to link the _new ask denom_ message to a gov proposal? Or it is automatically handled by the gov proposal execution?

File: x/ecocredit/spec/06_tests.md [hunk 6]
```diff
@@ -281,3 +269,185 @@ If a user tries to cancel credits and the user provides an invalid batch denomin
...
+- WHEN - user tries to add an ask denom
+- THEN - transaction is successful, ask denom is added
+
+If a user tries to add an ask denom and the user submits a governance proposal that is NOT approved, then the transaction fails and the ask denom is NOT added.
+
+- GIVEN - user provides a bid price using an ask denom that is NOT allowed
+- WHEN - user tries to add an ask denom
+- THEN - transaction fails, ask denom is NOT added
+
+If a user tries to add an ask denom and the user submits a transaction directly from their account, then the transaction fails and the ask denom is NOT added.
```
Comment (robert-zaremba) [inline x/ecocredit/spec/06_tests.md:449]: Let's use more simple language:
```suggestion
User can't add a new _ask denom_ directly, without creating (or linking... let's be more precise here) a gov proposal.
```

**Review by robert-zaremba — COMMENTED** [review]
Thanks for this great spec update.
Maybe we can change the language a bit for the acceptance tests. Instead:

> If a user ....

to

> User can / can't ...

If / else formula is already in given - when - then structure.

File: x/ecocredit/spec/06_tests.md (hunk 6 above)
Comment (ryanchristo) [reply x/ecocredit/spec/06_tests.md:449]: I think we should stick with the standard format for acceptance tests: https://openclassrooms.com/en/courses/4544611-write-agile-documentation-user-stories-acceptance-tests/4810081-write-acceptance-tests

Comment (ryanchristo) [reply x/ecocredit/spec/06_tests.md:449]: I don't think changing the language is necessary. I think we should follow the standard format for acceptance tests: https://openclassrooms.com/en/courses/4544611-write-agile-documentation-user-stories-acceptance-tests/4810081-write-acceptance-tests

File: x/ecocredit/spec/01_concepts.md (hunk 4 above)
Comment (ryanchristo) [reply x/ecocredit/spec/01_concepts.md:56]: Great. I can add the highlights. Maybe "One-Sided Orderbook" to be more descriptive?

File: x/ecocredit/spec/06_tests.md (hunk 5 above)
Comment (ryanchristo) [reply x/ecocredit/spec/06_tests.md:437]: The gov proposal executes the AllowAskDenom message. I can clarify.

**Comment by aaronc** [comment]
Excited to see these spec test cases. I wonder if we could make them executable with some go version of cucumber

**Comment by aaronc** [comment]
https://github.com/cucumber/godog

**Review by clevinson — APPROVED** [review]

**Comment by ryanchristo** [comment]
Opened #711 to further discuss and track updates to format and execution.

File: x/ecocredit/spec/06_tests.md (hunk 6 above)
Comment (robert-zaremba) [reply x/ecocredit/spec/06_tests.md]: if-then .... is just another way of saying the same thing using "given - when - then". 

The article says about the benefit of using "given - when -then" compared to other format, and it mentions the "if - else" for an exercise:

> Let's look at the first example from above:
> >    If the customer tries to withdraw $20 but her account balance is under $20, then do not allow the withdrawal.
>
> Writing that in Given-When-Then format will look like this:
> ...

File: x/ecocredit/spec/06_tests.md [hunk 7]
```diff
@@ -281,3 +269,185 @@ If a user tries to cancel credits and the user provides an invalid batch denomin
...
+- WHEN - user tries to create a sell order
+- THEN - transaction fails, sell order is NOT created
+
+## Update Sell Order
+
+If a user tries to update a sell order and the user provides a sell order id that does exist, then the transaction is successful and the sell order is updated.
+
+- GIVEN - user provides a sell order id that does exist
+- WHEN - user tries to update a sell order
+- THEN - transaction is successful, sell order is updated
```
Comment (robert-zaremba) [inline x/ecocredit/spec/06_tests.md:329]: I think  user should be able to update only his order.

**Pull request merged**